### PR TITLE
#167727729 User should be able to reset Password

### PR DESCRIPTION
**Pull request by MajemiteJames**
#### What does this PR do?
- Adds endpoint/feature so User can reset Password

#### Description
-Create endpoint to send email to user to reset Password

- Create endpoint to update user password

- write unit test for my implementation

- Document the API endpoint using swagger

#### How can this be manually tested
- Clone the repo
- Install dependencies, setup database and run `npm start:dev`
- Using postman or any http request tool, make a post request to `/api/v1/requests/password/reset to get an email notification to reset your password
-Clicking on the route it would send you to another route to update your password
- Test with postman or go to this url to test with swagger `http://localhost:3000/api/v1/docs`

#### Relevant Pivotal tracker stories
- [#167727729](https://www.pivotaltracker.com/story/show/167727736)

### CODE REVIEW

**Review by ladaposamuel — APPROVED** [review]
Looks good

File: src/controllers/userController.js [hunk 1]
```diff
@@ -157,11 +159,59 @@ const updateUserDetails = async (req, res) => {
...
+  * @returns {void} - no data returned
+*/
+const updatePassword = async (req, res) => {
+  try {
+    const { userId } = req.params;
+    const { password } = req.body;
+
+    const salt = bcrypt.genSaltSync(parseInt(process.env.SALT_ROUNDS, 10));
+
+    const newPassword = bcrypt.hashSync(password, salt);
```
Comment (tejiri4) [inline src/controllers/userController.js]: you dont need to encrypt the password a hooks does that

Comment (ladaposamuel) [reply src/controllers/userController.js]: Oops I missed that

Comment (Mcdavid95) [reply src/controllers/userController.js]: The hook is a beforeCreate hook, not sure it triggers on update

Comment (Mcdavid95) [reply src/controllers/userController.js]: @majemite you could check out the beforeUpdate hooks

**Comment by tejiri4** [comment]
Write  test and rebase

**Review by Mcdavid95 — APPROVED** [review]
Looks good to me, great job Chief😁

**Review by tejiri4 — CHANGES_REQUESTED** [review]
1. remove other field and leave email on your swagger doc request body when a user forget his/her password.

File: src/controllers/userController.js [hunk 2]
```diff
@@ -157,11 +160,60 @@ const updateUserDetails = async (req, res) => {
...
+  * @param {Object} res - response object from the server
+  *
+  * @returns {Object} - password reset link
+*/
+const resetPassword = async (req, res) => {
+  try {
+    const { email } = req.body;
+    const user = await findByEmail(email);
+    const token = generateVerificationToken({ id: user.id });
+    const link = `${process.env.BASE_URL}/update_password/${user.id}/${token}`;
```
Comment (ladaposamuel) [inline src/controllers/userController.js]: change to `BACKEND_BASE_URL`

**Review by ladaposamuel — CHANGES_REQUESTED** [review]

**Comment by ladaposamuel** [comment]
Rebase with the develop to solve conflicts

File: src/controllers/userController.js [hunk 3]
```diff
@@ -219,6 +217,58 @@ const socialAuth = async (req, res) => {
...
+*/
+const forgotPassword = async (req, res) => {
+  try {
+    const { email } = req.body;
+    const user = await findByEmail(email);
+    if (!user) {
+      const token = generateToken(email, '10m');
+      const link = `${process.env.FONTEND_BASE_URL}/reset/password/1/${token}`;
+      const message = createTemplate(resetPasswordMessage, link);
+      await sendMail(email, 'Reset Password', message);
```
Comment (tejiri4) [inline src/controllers/userController.js]: There no need to send an email if user dont exist just send a success response only

File: src/controllers/userController.js (hunk 2 above)
Comment (MajemiteJames) [reply src/controllers/userController.js]: done that

File: src/controllers/userController.js (hunk 3 above)
Comment (MajemiteJames) [reply src/controllers/userController.js]: changed to just success response

**Review by ladaposamuel — APPROVED** [review]
Looks good

File: src/controllers/userController.js [hunk 4]
```diff
@@ -221,6 +222,54 @@ const socialAuth = async (req, res) => {
...
+  * @param {Object} res - response object from the server
+  *
+  * @returns {Object} - password reset link
+*/
+const forgotPassword = async (req, res) => {
+  try {
+    const { email } = req.body;
+    const user = await findByEmail(email);
+    if (!user) {
+      return response(res, 200, 'success');
```
Comment (tejiri4) [inline src/controllers/userController.js]: send this message ```message: 'Check your mail to reset your password.``` in your response if no user

File: src/controllers/userController.js [hunk 5]
```diff
@@ -221,6 +222,54 @@ const socialAuth = async (req, res) => {
...
+    const { userId, token } = req.params;
+    const { password } = req.body;
+    const user = await User.findByPk(userId);
+    await verifyResetPasswordToken(token);
+    if (!user) return response(res, 404, 'error', { message: messages.userNotFound });
+    const options = { returning: true, where: { id: userId } };
+    const updatedUser = await update(User, { password }, options);
+    return response(res, 200, 'success', { message: 'Password updated successfully' }, updatedUser.lastName);
+  } catch (error) {
+    response(res, 500, 'error', { error: error.message });
```
Comment (ladaposamuel) [inline src/controllers/userController.js:270]: you are not returning this message, coveralls will miss this.

**Review by ladaposamuel — CHANGES_REQUESTED** [review]

File: src/utils/authHelper.js [hunk 6]
```diff
@@ -25,6 +25,15 @@ export const verifyToken = async (token) => {
   return decoded;
 };
 
+export const verifyResetPasswordToken = async (token) => {
+  const decoded = await jwt.verify(token, process.env.SECRET_KEY, (error) => {
+    if (error) {
+      return 'password reset link is invalid or has expired';
```
Comment (ladaposamuel) [inline src/utils/authHelper.js]: can you add this message to the global message helper file instead

**Review by ladaposamuel — CHANGES_REQUESTED** [review]

**Review by tejiri4 — APPROVED** [review]
LGTM

**Pull request merged**